rclc_lifecycle/src/rclc_lifecycle/rclc_lifecycle.c

```diff
@@ -265,3 +304,156 @@ rclc_lifecycle_execute_callback(
 
   return (*lifecycle_node->callbacks.fun_ptrs[transition_id])();
 }
+
+rcl_ret_t
+rclc_lifecycle_init_get_state_server(
+  rclc_lifecycle_node_t * lifecycle_node,
+  rclc_executor_t * executor)
+{
+  rclc_lifecycle_service_context_t context;
+  context.lifecycle_node = lifecycle_node;
+
+  rcl_ret_t rc = rclc_executor_add_service_with_context(
+    executor,
+    &lifecycle_node->state_machine->com_interface.srv_get_state,
+    &lifecycle_node->gs_req,
+    &lifecycle_node->gs_res,
+    rclc_lifecycle_get_state_callback,
+    &context);
+  if (rc != RCL_RET_OK) {
+    PRINT_RCLC_ERROR(main, rclc_executor_add_service_with_context);
+    return RCL_RET_ERROR;
+  }
+
+  return rc;
+}
+
+void
+rclc_lifecycle_get_state_callback(
+  const void * req,
+  void * res,
+  void * context)
+{
+  RCL_UNUSED(req);
+  lifecycle_msgs__srv__GetState_Response * res_in =
+    (lifecycle_msgs__srv__GetState_Response *) res;
+  rclc_lifecycle_service_context_t * context_in =
+    (rclc_lifecycle_service_context_t *) context;
+
+  rcl_lifecycle_state_machine_t * sm =
+    context_in->lifecycle_node->state_machine;
+
+  res_in->current_state.id = sm->current_state->id;
+  bool success = rosidl_runtime_c__String__assign(
+    &res_in->current_state.label,
+    sm->current_state->label);
+  if (!success) {
+    PRINT_RCLC_ERROR(
+      rclc_lifecycle_get_state_callback,
+      rosidl_runtime_c__String__assign);
+  }
+}
+
+rcl_ret_t
+rclc_lifecycle_init_get_available_states_server(
+  rclc_lifecycle_node_t * lifecycle_node,
+  rclc_executor_t * executor)
+{
+  rclc_lifecycle_service_context_t context;
+  context.lifecycle_node = lifecycle_node;
+
+  rcl_ret_t rc = rclc_executor_add_service_with_context(
+    executor,
+    &lifecycle_node->state_machine->com_interface.srv_get_available_states,
+    &lifecycle_node->gas_req,
+    &lifecycle_node->gas_res,
+    rclc_lifecycle_get_available_states_callback,
+    &context);
+  if (rc != RCL_RET_OK) {
+    PRINT_RCLC_ERROR(main, rclc_executor_add_service_with_context);
+    return RCL_RET_ERROR;
+  }
+
+  return rc;
+}
+
+void
```

Comment:
same comment as above - you don't want to return `rcl_ret_t`?

Reply:
These functions implement a `rclc_service_callback_with_context_t`, which is 
`typedef void (* rclc_service_callback_with_context_t)(const void *, void *, void *);`

Do you think this would benefit from a return value? Would the rcl server be able to handle it approprietly, @JanStaschulat?

Reply:
Okay, understood. Missed that information. Then `void ` is okay.